Source/WebCore/accessibility/AXTextMarker.cpp

```diff
@@ -732,20 +740,38 @@ const AXTextRuns* AXTextMarker::runs() const
     return object ? object->textRuns() : nullptr;
 }
 
-AXTextMarker AXTextMarker::findMarker(AXDirection direction, std::optional<AXID> stopAtID) const
+AXTextMarker AXTextMarker::findMarker(AXDirection direction, CoalesceObjectBreaks coalesceObjectBreaks, IgnoreBRs ignoreBRs, std::optional<AXID> stopAtID) const
 {
+    // This method has two boolean options:
+    // - coalesceObjectBreaks: Mimics behavior from textMarkerDataForNextCharacterOffset, where we skip nodes
+    //   that have the same visual position (i.e., there is 0 length between them). When false, we traverse all
+    //   possible text markers (which is important for searching)
+    // - ignoreBRs: In most cases, we want to skip <br> tags when not in an editable context. This is not true,
+    //   for example, when computing text marker indexes.
+
     if (!isValid())
         return { };
     if (!isInTextRun())
-        return toTextRunMarker().findMarker(direction, stopAtID);
+        return toTextRunMarker().findMarker(direction, coalesceObjectBreaks, ignoreBRs, stopAtID);
+
+    RefPtr object = isolatedObject();
 
-    if ((direction == AXDirection::Next && offset() < runs()->totalLength()) || (direction == AXDirection::Previous && offset() > 0))
-        return { treeID(), objectID(), direction == AXDirection::Next ? offset() + 1 : offset() - 1 };
+    // If the BR isn't in an editable ancestor, we shouldn't be including it (in most cases of findMarker).
+    bool shouldSkipBR = ignoreBRs == IgnoreBRs::Yes && object && object->roleValue() == AccessibilityRole::LineBreak && !object->editableAncestor();
+    bool isWithinRunBounds = ((direction == AXDirection::Next && offset() < runs()->totalLength()) || (direction == AXDirection::Previous && offset()));
+    if (!shouldSkipBR && isWithinRunBounds)
+        return AXTextMarker { treeID(), objectID(), direction == AXDirection::Next ? offset() + 1 : offset() - 1 };
 
     // offset() pointed to the last character in the given object's runs, so let's traverse to find the next object with runs.
     if (RefPtr object = findObjectWithRuns(*isolatedObject(), direction, stopAtID)) {
         RELEASE_ASSERT(direction == AXDirection::Next ? object->textRuns()->runLength(0) : object->textRuns()->lastRunLength());
-        return { object->treeID(), object->objectID(), direction == AXDirection::Next ? 0 : object->textRuns()->lastRunLength() };
+
+        // The startingOffset is used to advance one position farther when we are coalescing object breaks and skipping positions.
+        int startingOffset = 0;
+        if ((coalesceObjectBreaks == CoalesceObjectBreaks::Yes || shouldSkipBR) && !isolatedObject()->shouldEmitNewlinesBeforeAndAfterNode())
+            startingOffset = 1;
+
+        return AXTextMarker { object->treeID(), object->objectID(), direction == AXDirection::Next ? startingOffset : object->textRuns()->lastRunLength() - startingOffset };
```

Comment:
Is it possible for `object->textRuns()->lastRunLength()` to be 0 and `startingOffset` to be 1, thus making this subtraction wrap backwards?

Reply:
Hmm, this didn't seem possible from my understanding, but please correct me if I'm wrong. `object` here comes from `nextObjectWithRuns`, which only returns the next object that satisfies `hasTextRuns()`:
```c++
bool hasTextRuns() final { return textRuns().size(); }
```

But, is it possible for `AXTextRuns` to have a run with a text length of 0? If so, then yes, we should guard against that.

Reply:
Actually, this should never be possible. A few lines up, we assert that `lastRunLength` != 0:
```c++
        RELEASE_ASSERT(direction == AXDirection::Next ? object->textRuns()->runLength(0) : object->textRuns()->lastRunLength());
```